### PR TITLE
[Core][Hotfix] Retake #2673. Filling buffer on assign processes

### PR DESCRIPTION
**Pull request by loumalouomega**
This retakes #2673 in order to fill the buffer. Preserves current behaviour by default. Modifies some C++ files necessaries in order to set values in custom buffer steps

### CODE REVIEW

File: kratos/python/add_utilities_to_python.cpp [hunk 1]
```diff
@@ -74,6 +74,77 @@ void SetOnProcessInfo(
     rCurrentProcessInfo[TABLE_UTILITY] = pTable;
 }
 
+double RotateAndCallFunctionCurrentConfiguration(
+    PythonGenericFunctionUtility& rPythonGenericFunctionUtility,
+    const double x,
+    const double y,
+    const double z,
```
Comment (pooyan-dadvand) [inline kratos/python/add_utilities_to_python.cpp]: Wouldn't it be better to use a Point or array_1d<double,3> instead of coordinates separately?

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: Yes..., but the initial implementation was with X, Y, Z and to avoid break things I left it that way

File: kratos/python/add_utilities_to_python.cpp [hunk 2]
```diff
@@ -285,6 +356,81 @@ void CopyModelPartNodalVarToNonHistoricalVarWithDestination(
...
+    const double Value,
+    NodesContainerType& rNodes,
+    const IndexType BuffStep
+    )
+{
+    rVariableUtils.SetScalarVar(rVariable, Value, rNodes, BuffStep);
+}
+
+template<class TVarType>
+void SetScalarVarForFlagWithoutBufferNotCustomCheck(
```
Comment (pooyan-dadvand) [inline kratos/python/add_utilities_to_python.cpp]: What is the name means?

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: By default the flag is set as true, but with this you can set as true or false

Comment (pooyan-dadvand) [reply kratos/python/add_utilities_to_python.cpp]: So wouldn't be better to have it as a default value in the other one?

File: kratos/python/add_utilities_to_python.cpp (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/python/add_utilities_to_python.cpp]: I see....

File: kratos/python/add_utilities_to_python.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: Yes, the problem is that these method gave errors when exporting to python and later execute them

Comment (pooyan-dadvand) [reply kratos/python/add_utilities_to_python.cpp]: Why? Would you please describe more?

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: At least what I saw, pybind may give problem when exporting to python methods with default values. When you try to execute them it says that the arguments are wrong

Comment (pooyan-dadvand) [reply kratos/python/add_utilities_to_python.cpp]: I see, but then you may use two different helper functions like overloaded methods. Isn't it?

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: It is what I am doing, but with different names

Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: just FYI: pybind cannot extract defaults from c++, but they can still be specified (separately)
See e.g. https://github.com/KratosMultiphysics/Kratos/blob/a1ca9e7d0fd006539f9605332a8a9444ad085cba/applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp#L358

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: Done, thanks

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: @pooyan-dadvand I removed those duplicated methods

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
Ping

File: kratos/python/add_utilities_to_python.cpp [hunk 3]
```diff
@@ -431,16 +509,16 @@ void AddUtilitiesToPython(pybind11::module &m)
...
-        .def("SetScalarVar", &VariableUtils::SetScalarVar<VariableComponent<VectorComponentAdaptor<array_1d<double, 4>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVar<VariableComponent<VectorComponentAdaptor<array_1d<double, 6>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVar<VariableComponent<VectorComponentAdaptor<array_1d<double, 9>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVarForFlag<Variable<double>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVarForFlag<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVarForFlag<VariableComponent<VectorComponentAdaptor<array_1d<double, 4>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVarForFlag<VariableComponent<VectorComponentAdaptor<array_1d<double, 6>>>>)
-        .def("SetScalarVar", &VariableUtils::SetScalarVarForFlag<VariableComponent<VectorComponentAdaptor<array_1d<double, 9>>>>)
+        .def("SetScalarVar", &VariableUtils::SetScalarVar<Variable<double>>, py::arg("variable"), py::arg("value"), py::arg("nodes"), py::arg("buffer_step")=0)
```
Comment (RiccardoRossi) [inline kratos/python/add_utilities_to_python.cpp]: what does this modification mean?

Comment (RiccardoRossi) [reply kratos/python/add_utilities_to_python.cpp]: i am referring to the syntax employed

Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: this is needed when using default args, as he does for `buffer_step`

Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp]: @philbucher already answered

File: kratos/python_scripts/assign_scalar_variable_process.py [hunk 4]
```diff
@@ -40,6 +37,7 @@ def __init__(self, Model, settings ):
             "variable_name"   : "SPECIFY_VARIABLE_NAME",
             "interval"        : [0.0, 1e30],
             "constrained"     : true,
+            "fill_buffer"     : true,
```
Comment (pooyan-dadvand) [inline kratos/python_scripts/assign_scalar_variable_process.py]: The default value is changing the current behavior. Isn't it?

**Comment by loumalouomega** [comment]
I don't remember right now, but if currently the buffer is not filled I
would say it changes the behaviour.

**Comment by loumalouomega** [comment]
Should I fix this?

**Comment by pooyan-dadvand** [comment]
To be discussed in @KratosMultiphysics/technical-committee

File: kratos/python/add_variable_utils_to_python.cpp [hunk 5]
```diff
@@ -218,29 +224,29 @@ void AddVariableUtilsToPython(pybind11::module &m)
         .def("CopyModelPartElementalVar", &VariableUtils::CopyModelPartElementalVar<Variable<Quaternion<double>>>)
         .def("CopyModelPartElementalVar", &VariableUtils::CopyModelPartElementalVar<Variable<Vector>>)
         .def("CopyModelPartElementalVar", &VariableUtils::CopyModelPartElementalVar<Variable<Matrix>>)
-        .def("SetVectorVar", VariableUtilsSetVariable<array_1d<double,3>>)
-        .def("SetVectorVar", VariableUtilsSetVariableForFlag<array_1d<double, 3>>)
-        .def("SetScalarVar", VariableUtilsSetVariable<double>)
-        .def("SetScalarVar", VariableUtilsSetVariableForFlag<double>)
+        .def("SetVectorVar", VariableUtilsSetVariable<array_1d<double,3>>, py::arg("variable"), py::arg("value"), py::arg("nodes"), py::arg("buffer_step")=0)
```
Comment (RiccardoRossi) [inline kratos/python/add_variable_utils_to_python.cpp]: i am not convinced on the use of py::arg, as from my understanding it uses under the hood kwargs.

so
1 - it has a performance penalty
2 - it is not compatible with current usage

File: kratos/python/add_other_utilities_to_python.cpp [hunk 6]
```diff
@@ -62,6 +62,77 @@ void SetOnProcessInfo(
     rCurrentProcessInfo[TABLE_UTILITY] = pTable;
 }
 
+double RotateAndCallFunctionCurrentConfiguration(
```
Comment (RiccardoRossi) [inline kratos/python/add_other_utilities_to_python.cpp:65]: as i understand those functions (which are surely interesting) do not really belong in this PR

Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp:65]: I will check

File: kratos/python/add_variable_utils_to_python.cpp (hunk 5 above)
Comment (loumalouomega) [reply kratos/python/add_variable_utils_to_python.cpp]: I did it at that moment as @philbucher, it can be replaced with the lambdas (or updating to C++14 :P )

Comment (philbucher) [reply kratos/python/add_variable_utils_to_python.cpp]: this has nothing to do with lambdas, @RiccardoRossi s concern is with the `py::args` being problematic (args vs kwargs) (I also don't yet know abt this :D)

File: kratos/python/add_other_utilities_to_python.cpp (hunk 6 above)
Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp:65]: It is used

**Review by philbucher — COMMENTED** [review]
@KratosMultiphysics/technical-committee took a look at this and we don't understand the changes in the py callback utility. Please separate this PR into two:
1. Changes in the Callback utility. Please explain what was missing (we checked master and it seems like all the required functionalities are already there)
2. The fill buffer changes in the Process

**Comment by loumalouomega** [comment]
I still think the idea of filling the buffer is important, but it requires a whole new PR
